pkg/activator/util/rewinder.go

```diff
@@ -16,44 +16,58 @@ package util
 import (
 	"bytes"
 	"io"
-	"io/ioutil"
 	"sync"
 )
 
 type rewinder struct {
 	sync.Mutex
-	rc io.ReadCloser
-	rs io.ReadSeeker
+	original io.ReadCloser
+	current  io.Reader
+	next     io.ReadWriter
+	eof      bool
 }
 
-// rewinder wraps a single-use `ReadCloser` into a `ReadCloser` that can be read multiple times
+// rewinder wraps a single-use `ReadCloser` into a `ReadCloser`
+// that can be read multiple times
 func NewRewinder(rc io.ReadCloser) io.ReadCloser {
-	return &rewinder{rc: rc}
+	r := &rewinder{original: rc}
+
+	r.next = new(bytes.Buffer)
+	r.current = r.original
+
+	return r
 }
 
 func (r *rewinder) Read(b []byte) (int, error) {
 	r.Lock()
 	defer r.Unlock()
-	// On the first `Read()`, the contents of `rc` is read into a buffer `rs`.
-	// This buffer is used for all subsequent reads
-	if r.rs == nil {
-		buf, err := ioutil.ReadAll(r.rc)
-		if err != nil {
-			return 0, err
-		}
-		r.rc.Close()
-
-		r.rs = bytes.NewReader(buf)
+
+	// Buffer everything we read
+	n, err := io.TeeReader(r.current, r.next).Read(b)
+
+	// Keep track of when we reach the end
+	if err == io.EOF {
+		r.eof = true
 	}
 
-	return r.rs.Read(b)
+	return n, err
 }
 
 func (r *rewinder) Close() error {
 	r.Lock()
 	defer r.Unlock()
-	// Rewind the buffer on `Close()` for the next call to `Read`
-	r.rs.Seek(0, io.SeekStart)
+
+	// Close the original ReadCloser if we have read it to the end
+	if r.eof && r.original != nil {
+		r.original.Close()
+		r.original = nil
+	}
+
+	// Rewind back to the beginning
+	start := io.MultiReader(r.next, r.current)
+
+	r.next = new(bytes.Buffer)
+	r.current = start
```

Comment:
This is an awesome enhancement of that bit of code. If I understood this correctly, we'll no longer necessarily need to read everything in memory once but can probably get away with only reading a few bytes, because the 503's should arrive based on headers already.

As future work: We should measure this and see how much this reads in memory in practice. Maybe we can even drop the maximum content-length enforced by the activator.

Good one 🎉

Reply:
>  If I understood this correctly, we'll no longer necessarily need to read everything in memory once but can probably get away with only reading a few bytes, because the 503's should arrive based on headers already.

We eventually _have_ to read everything to memory, because once the 503's are cleared, the request has to be proxied to the revision. This enhancement is necessary for supporting gRPC streams; without it, the activator buffers the request body forever. Ideally, we know ahead of time when Istio will stop returning 503s and we don't have to use the rewinder at all. 😄